src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+      return $ TabTy (b':>IxType iTy' d') resultTy'
+  RecordTy  elems -> RecordTy  <$> traverserseFieldRowElemTypes (f TypeParam TyKind) elems
+  VariantTy (Ext elems Nothing) -> do
+    elems' <- traverse (f TypeParam TyKind) elems
+    return $ VariantTy $ Ext elems' Nothing
+  TC tc -> TC <$> case tc of
+    BaseType b -> return $ BaseType b
+    ProdType tys -> ProdType <$> forM tys \t -> f TypeParam TyKind t
+    SumType  tys -> SumType  <$> forM tys \t -> f TypeParam TyKind t
+    Nat -> return Nat
+    Fin n -> Fin <$> f DataParam NatTy n
+    RefType _ _ -> error "not implemented" -- how should we handle the ParamRole for the heap parameter?
+    TypeKind         -> return TypeKind
+    EffectRowKind    -> return EffectRowKind
+    LabeledRowKindTC -> return LabeledRowKindTC
+    LabelType        -> return LabelType
+  _ -> error $ "Not implemented: " ++ pprint ty
+{-# INLINE traverseTyParams #-}
+
+traverseRoleBinders
+  :: forall m n n'. EnvReader m
+  => (forall l . DExt n l => ParamRole -> Type l -> Atom l -> m l (Atom l))
+  ->  Nest RoleBinder n n' -> [Atom n] -> m n [Atom n]
+traverseRoleBinders f allBinders allParams =
+  runSubstReaderT idSubst $ go allBinders allParams
+  where
+    go :: forall i i'. Nest RoleBinder i i' -> [Atom n]
+       -> SubstReaderT AtomSubstVal m i n [Atom n]
+    go Empty [] = return []
+    go (Nest (RoleBinder b ty role) bs) (param:params) = do
+      ty' <- substM ty
+      Distinct <- getDistinct
+      param' <- liftSubstReaderT $ f role ty' param
+      params'' <- extendSubst (b@>SubstVal param') $ go bs params
+      return $ param' : params''
+    go _ _ = error "zip error"
+{-# INLINE traverseRoleBinders #-}
+
+
+traverserseFieldRowElemTypes :: Monad m => (Type n -> m (Type n)) -> FieldRowElems n -> m (FieldRowElems n)
+traverserseFieldRowElemTypes f els = fieldRowElemsFromList <$> traverse checkElem elemList
+  where
+    elemList = fromFieldRowElems els
+    checkElem = \case
+      StaticFields items -> StaticFields <$> traverse f items
+      DynField _ _ -> error "shouldn't have dynamic fields post-simplification"
+      DynFields _  -> error "shouldn't have dynamic fields post-simplification"
+
+-- traverseLabeledRow
+--   :: Monad m
+--   => (Type n -> m (Type n))
+--   -> LabeledItems (Type n)
+--   -> m (LabeledItems (Type n))
+-- traverseLabeledRow f (items rest) = undefined
+-- {-# INLINE traverseLabeledRow #-}
+
+-- traverseLabeledRow (Ext items rest) = do
+--   mapM_ (|: TyKind) items
+--   forM_ rest \name -> do
+--     name' <- lookupSubstM name
+--     ty <- atomBindingType <$> lookupEnv name'
+--     checkAlphaEq LabeledRowKind ty
+
+
+-- checkLabeledRow :: Typer m => ExtLabeledItems (Type i) (AtomName i) -> m i o ()
+-- checkLabeledRow (Ext items rest) = do
+--   mapM_ (|: TyKind) items
+--   forM_ rest \name -> do
+--     name' <- lookupSubstM name
+--     ty <- atomBindingType <$> lookupEnv name'
+--     checkAlphaEq LabeledRowKind ty
```

Comment:
Why are those commented out? Should they be implemented? Or removed?

Reply:
Removed. Thanks for catching that.